File: datastructures/hashmap.py
class HashMap:
    def __init__(self, capacity=16):
        self.capacity = capacity
        self.size = 0
        self.buckets = [[] for _ in range(self.capacity)]  # each bucket is a list of (key, value) pairs

    def _hash(self, key):
        return hash(key) % self.capacity

    # Insert / Update - O(1) average
    def put(self, key, value):
        idx = self._hash(key)
        bucket = self.buckets[idx]
        for i, (k, v) in enumerate(bucket):
            if k == key:
                bucket[i] = (key, value)    # update existing key
                return
        bucket.append((key, value))         # insert new key
        self.size += 1

        # Resize if load factor exceeds 0.75
        if self.size / self.capacity > 0.75:
            self._resize()

    # Get value by key - O(1) average
    def get(self, key):
        idx = self._hash(key)
        bucket = self.buckets[idx]
        for k, v in bucket:
            if k == key:
                return v
        raise KeyError(f"Key '{key}' not found")

    # Delete by key - O(1) average
    def delete(self, key):
        idx = self._hash(key)
        bucket = self.buckets[idx]
        for i, (k, v) in enumerate(bucket):
            if k == key:
                bucket.pop(i)
                self.size -= 1
                return
        raise KeyError(f"Key '{key}' not found")

    # Check key existence - O(1) average
    def contains(self, key):
        idx = self._hash(key)
        bucket = self.buckets[idx]
        return any(k == key for k, v in bucket)

    # Resize when load factor exceeded
    def _resize(self):
        old_buckets = self.buckets
        self.capacity *= 2
        self.buckets = [[] for _ in range(self.capacity)]
        self.size = 0
        for bucket in old_buckets:
            for k, v in bucket:
                self.put(k, v)

    def display(self):
        for i, bucket in enumerate(self.buckets):
            if bucket:
                print(f"  bucket[{i}] -> {bucket}")


hm = HashMap()
hm.put("name", "Alice")
hm.put("age", 30)
hm.put("city", "Bengaluru")
hm.put("age", 31)               # update existing key

print(hm.get("name"))           # Alice
print(hm.get("age"))            # 31
print(hm.contains("city"))      # True
print(hm.contains("salary"))    # False

hm.delete("city")
print(hm.contains("city"))      # False

hm.display()

File: datastructures/HashMap.java
import java.util.ArrayList;
import java.util.LinkedList;

public class HashMap {

    // Each entry holds a key-value pair
    static class Entry {
        String key;
        int value;

        Entry(String key, int value) {
            this.key = key;
            this.value = value;
        }
    }

    private int capacity;
    private int size;
    private LinkedList<Entry>[] buckets;
    private static final double LOAD_FACTOR = 0.75;

    @SuppressWarnings("unchecked")
    public HashMap(int capacity) {
        this.capacity = capacity;
        this.size = 0;
        this.buckets = new LinkedList[capacity];
        for (int i = 0; i < capacity; i++) {
            buckets[i] = new LinkedList<>();
        }
    }

    public HashMap() {
        this(16);
    }

    private int hash(String key) {
        return Math.abs(key.hashCode()) % capacity;
    }

    // Insert / Update - O(1) average
    public void put(String key, int value) {
        int idx = hash(key);
        LinkedList<Entry> bucket = buckets[idx];

        for (Entry entry : bucket) {
            if (entry.key.equals(key)) {
                entry.value = value;    // update existing
                return;
            }
        }

        bucket.add(new Entry(key, value));  // insert new
        size++;

        if ((double) size / capacity > LOAD_FACTOR) {
            resize();
        }
    }

    // Get value by key - O(1) average
    public int get(String key) {
        int idx = hash(key);
        for (Entry entry : buckets[idx]) {
            if (entry.key.equals(key)) return entry.value;
        }
        throw new RuntimeException("Key '" + key + "' not found");
    }

    // Delete by key - O(1) average
    public void delete(String key) {
        int idx = hash(key);
        LinkedList<Entry> bucket = buckets[idx];
        for (Entry entry : bucket) {
            if (entry.key.equals(key)) {
                bucket.remove(entry);
                size--;
                return;
            }
        }
        throw new RuntimeException("Key '" + key + "' not found");
    }

    // Check key existence - O(1) average
    public boolean contains(String key) {
        int idx = hash(key);
        for (Entry entry : buckets[idx]) {
            if (entry.key.equals(key)) return true;
        }
        return false;
    }

    // Resize when load factor exceeded
    @SuppressWarnings("unchecked")
    private void resize() {
        int newCapacity = capacity * 2;
        LinkedList<Entry>[] newBuckets = new LinkedList[newCapacity];
        for (int i = 0; i < newCapacity; i++) {
            newBuckets[i] = new LinkedList<>();
        }

        for (LinkedList<Entry> bucket : buckets) {
            for (Entry entry : bucket) {
                int newIdx = Math.abs(entry.key.hashCode()) % newCapacity;
                newBuckets[newIdx].add(entry);
            }
        }

        buckets = newBuckets;
        capacity = newCapacity;
    }

    public int size() { return size; }

    public void display() {
        for (int i = 0; i < capacity; i++) {
            if (!buckets[i].isEmpty()) {
                System.out.print("  bucket[" + i + "] -> ");
                for (Entry entry : buckets[i]) {
                    System.out.print("{" + entry.key + ": " + entry.value + "} ");
                }
                System.out.println();
            }
        }
    }

    public static void main(String[] args) {
        HashMap hm = new HashMap();
        hm.put("name", 1);
        hm.put("age", 30);
        hm.put("city", 99);
        hm.put("age", 31);              // update existing key

        System.out.println(hm.get("age"));      // 31
        System.out.println(hm.contains("city")); // true

        hm.delete("city");
        System.out.println(hm.contains("city")); // false

        hm.display();
    }
}
